File: packages/domain/src/color.ts
export type OklchColor = {
  lightness: number;
  chroma: number;
  hue: number;
};

type RgbColor = {
  red: number;
  green: number;
  blue: number;
};

const HEX_COLOR_PATTERN = /^#([0-9a-f]{3}|[0-9a-f]{6})$/i;
const OKLCH_COLOR_PATTERN = /^oklch\(\s*(\d*\.?\d+)\s+(\d*\.?\d+)\s+(\d*\.?\d+)\s*\)$/i;

export function parseOklchColor(value: string): OklchColor | null {
  const match = value.match(OKLCH_COLOR_PATTERN);
  if (!match) return null;
  const lightness = Number(match[1]);
  const chroma = Number(match[2]);
  const hue = Number(match[3]);
  if (!Number.isFinite(lightness) || lightness < 0 || lightness > 1) return null;
  if (!Number.isFinite(chroma) || chroma < 0 || chroma > 0.4) return null;
  if (!Number.isFinite(hue) || hue < 0 || hue >= 360) return null;
  return { lightness, chroma, hue };
}

export function hexToOklch(value: string): string | null {
  const rgb = rgbOfHex(value);
  if (!rgb) return null;
  const red = srgbToLinear(rgb.red);
  const green = srgbToLinear(rgb.green);
  const blue = srgbToLinear(rgb.blue);
  const l = Math.cbrt(0.4122214708 * red + 0.5363325363 * green + 0.0514459929 * blue);
  const m = Math.cbrt(0.2119034982 * red + 0.6806995451 * green + 0.1073969566 * blue);
  const s = Math.cbrt(0.0883024619 * red + 0.2817188376 * green + 0.6299787005 * blue);
  const lightness = 0.2104542553 * l + 0.793617785 * m - 0.0040720468 * s;
  const a = 1.9779984951 * l - 2.428592205 * m + 0.4505937099 * s;
  const b = 0.0259040371 * l + 0.7827717662 * m - 0.808675766 * s;
  const chroma = Math.hypot(a, b);
  const hue = chroma < 0.0005 ? 0 : (Math.atan2(b, a) * 180 / Math.PI + 360) % 360;
  return formatOklchColor({ lightness, chroma, hue });
}

export function oklchToHex(value: string): string | null {
  const color = parseOklchColor(value);
  if (!color) return null;
  const rgb = rgbOfOklch(color);
  return `#${[rgb.red, rgb.green, rgb.blue]
    .map((channel) => Math.round(clamp(channel) * 255).toString(16).padStart(2, "0"))
    .join("")}`;
}

export function normalizeColorToOklch(value: string): string | null {
  const parsed = parseOklchColor(value);
  if (parsed) return formatOklchColor(parsed);
  return hexToOklch(value);
}

export function relativeLuminanceOfColor(value: string): number | null {
  const parsed = parseOklchColor(value);
  const rgb = parsed ? rgbOfOklch(parsed) : rgbOfHex(value);
  if (!rgb) return null;
  return 0.2126 * srgbToLinear(rgb.red) + 0.7152 * srgbToLinear(rgb.green) + 0.0722 * srgbToLinear(rgb.blue);
}

function formatOklchColor(color: OklchColor) {
  return `oklch(${formatChannel(color.lightness)} ${formatChannel(color.chroma)} ${formatChannel(color.hue)})`;
}

function formatChannel(value: number) {
  const formatted = value.toFixed(3).replace(/\.?0+$/, "");
  return formatted === "-0" ? "0" : formatted;
}

function rgbOfHex(value: string): RgbColor | null {
  const match = value.match(HEX_COLOR_PATTERN);
  if (!match) return null;
  const normalized = match[1].length === 3
    ? [...match[1]].map((digit) => digit.repeat(2)).join("")
    : match[1];
  return {
    red: Number.parseInt(normalized.slice(0, 2), 16) / 255,
    green: Number.parseInt(normalized.slice(2, 4), 16) / 255,
    blue: Number.parseInt(normalized.slice(4, 6), 16) / 255,
  };
}

function rgbOfOklch({ lightness, chroma, hue }: OklchColor): RgbColor {
  const radians = hue * Math.PI / 180;
  const a = chroma * Math.cos(radians);
  const b = chroma * Math.sin(radians);
  const l = (lightness + 0.3963377774 * a + 0.2158037573 * b) ** 3;
  const m = (lightness - 0.1055613458 * a - 0.0638541728 * b) ** 3;
  const s = (lightness - 0.0894841775 * a - 1.291485548 * b) ** 3;
  return {
    red: linearToSrgb(4.0767416621 * l - 3.3077115913 * m + 0.2309699292 * s),
    green: linearToSrgb(-1.2684380046 * l + 2.6097574011 * m - 0.3413193965 * s),
    blue: linearToSrgb(-0.0041960863 * l - 0.7034186147 * m + 1.707614701 * s),
  };
}

function srgbToLinear(channel: number) {
  return channel <= 0.04045 ? channel / 12.92 : ((channel + 0.055) / 1.055) ** 2.4;
}

function linearToSrgb(channel: number) {
  return channel <= 0.0031308 ? 12.92 * channel : 1.055 * channel ** (1 / 2.4) - 0.055;
}

function clamp(channel: number) {
  return Math.min(1, Math.max(0, channel));
}

File: crates/mono-api/src/color.rs
// Ported from packages/domain/src/color.ts. Only as much as is needed for todo/ledger/calendar label color validation:
// normalize_color_to_oklch (parses oklch, or converts hex→oklch). The rest (oklch→hex etc.) can be added when needed.

struct Oklch {
    lightness: f64,
    chroma: f64,
    hue: f64,
}

/// Normalizes oklch(...) or a 6/3-digit hex into an `oklch(L C H)` string. None on a malformed input.
pub fn normalize_color_to_oklch(value: &str) -> Option<String> {
    if let Some(parsed) = parse_oklch(value) {
        return Some(format_oklch(&parsed));
    }
    hex_to_oklch(value)
}

fn parse_oklch(value: &str) -> Option<Oklch> {
    // Numbers, whitespace, and parens are unaffected by lowercasing — only the prefix is matched case-insensitively.
    let lowered = value.trim().to_ascii_lowercase();
    let inner = lowered.strip_prefix("oklch(")?.strip_suffix(')')?;
    let parts: Vec<f64> = inner
        .split_whitespace()
        .map(|token| token.parse::<f64>())
        .collect::<Result<_, _>>()
        .ok()?;
    if parts.len() != 3 {
        return None;
    }
    let (lightness, chroma, hue) = (parts[0], parts[1], parts[2]);
    if !lightness.is_finite() || !(0.0..=1.0).contains(&lightness) {
        return None;
    }
    if !chroma.is_finite() || !(0.0..=0.4).contains(&chroma) {
        return None;
    }
    if !hue.is_finite() || !(0.0..360.0).contains(&hue) {
        return None;
    }
    Some(Oklch { lightness, chroma, hue })
}

fn hex_to_oklch(value: &str) -> Option<String> {
    let (r, g, b) = rgb_of_hex(value)?;
    let red = srgb_to_linear(r);
    let green = srgb_to_linear(g);
    let blue = srgb_to_linear(b);
    let l = (0.412_221_470_8 * red + 0.536_332_536_3 * green + 0.051_445_992_9 * blue).cbrt();
    let m = (0.211_903_498_2 * red + 0.680_699_545_1 * green + 0.107_396_956_6 * blue).cbrt();
    let s = (0.088_302_461_9 * red + 0.281_718_837_6 * green + 0.629_978_700_5 * blue).cbrt();
    let lightness = 0.210_454_255_3 * l + 0.793_617_785 * m - 0.004_072_046_8 * s;
    let a = 1.977_998_495_1 * l - 2.428_592_205 * m + 0.450_593_709_9 * s;
    let b_comp = 0.025_904_037_1 * l + 0.782_771_766_2 * m - 0.808_675_766 * s;
    let chroma = a.hypot(b_comp);
    let hue = if chroma < 0.0005 {
        0.0
    } else {
        (b_comp.atan2(a) * 180.0 / std::f64::consts::PI + 360.0) % 360.0
    };
    Some(format_oklch(&Oklch { lightness, chroma, hue }))
}

fn format_oklch(color: &Oklch) -> String {
    format!(
        "oklch({} {} {})",
        format_channel(color.lightness),
        format_channel(color.chroma),
        format_channel(color.hue)
    )
}

// Equivalent to JS `value.toFixed(3).replace(/\.?0+$/, "")`: rounds to 3 decimals then strips trailing zeros/dot, and "-0"→"0".
fn format_channel(value: f64) -> String {
    let fixed = format!("{value:.3}");
    let trimmed = if fixed.contains('.') {
        fixed.trim_end_matches('0').trim_end_matches('.')
    } else {
        &fixed
    };
    if trimmed == "-0" {
        "0".to_string()
    } else {
        trimmed.to_string()
    }
}

fn rgb_of_hex(value: &str) -> Option<(f64, f64, f64)> {
    let hex = value.strip_prefix('#')?;
    if !hex.chars().all(|c| c.is_ascii_hexdigit()) {
        return None;
    }
    let full = match hex.len() {
        3 => hex.chars().flat_map(|c| [c, c]).collect::<String>(),
        6 => hex.to_string(),
        _ => return None,
    };
    let channel = |slice: &str| u8::from_str_radix(slice, 16).ok().map(|v| v as f64 / 255.0);
    Some((channel(&full[0..2])?, channel(&full[2..4])?, channel(&full[4..6])?))
}

fn srgb_to_linear(channel: f64) -> f64 {
    if channel <= 0.04045 {
        channel / 12.92
    } else {
        ((channel + 0.055) / 1.055).powf(2.4)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn passes_through_normalized_oklch() {
        assert_eq!(
            normalize_color_to_oklch("oklch(0.645 0.009 106.643)").as_deref(),
            Some("oklch(0.645 0.009 106.643)")
        );
    }

    #[test]
    fn trims_trailing_zeros() {
        assert_eq!(normalize_color_to_oklch("oklch(0.500 0.000 0.000)").as_deref(), Some("oklch(0.5 0 0)"));
    }

    #[test]
    fn converts_black_hex() {
        assert_eq!(normalize_color_to_oklch("#000000").as_deref(), Some("oklch(0 0 0)"));
    }

    #[test]
    fn converts_short_hex_and_matches_full() {
        assert_eq!(normalize_color_to_oklch("#fff"), normalize_color_to_oklch("#ffffff"));
    }

    #[test]
    fn converts_arbitrary_hex() {
        // The reference value from color.ts hexToOklch("#b03a55").
        let out = normalize_color_to_oklch("#b03a55").unwrap();
        assert!(out.starts_with("oklch(0.5"), "got {out}");
    }

    #[test]
    fn rejects_garbage() {
        assert_eq!(normalize_color_to_oklch("red"), None);
        assert_eq!(normalize_color_to_oklch("#12"), None);
        assert_eq!(normalize_color_to_oklch("oklch(2 0 0)"), None);
        assert_eq!(normalize_color_to_oklch("rgb(0,0,0)"), None);
    }
}
